### PR TITLE
test/region_syncer: make TestPrepareCheckerWithTransferLeader more stable

### LINKED ISSUE #8351: TestPrepareCheckerWithTransferLeader is unstable
### Which jobs are failing
```
--- FAIL: TestPrepareCheckerWithTransferLeader (38.21s)
    region_syncer_test.go:294: 
        	Error Trace:	/home/runner/work/pd/pd/tests/server/region_syncer/region_syncer_test.go:294
        	Error:      	Not equal: 
        	            	expected: "pd1"
        	            	actual  : "pd2"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-pd1
        	            	+pd2
        	Test:       	TestPrepareCheckerWithTransferLeader
FAIL
```

### CI link

https://github.com/tikv/pd/actions/runs/9737924850/job/26924893445

### PR DESCRIPTION
**Pull request by HuSharp**
### What problem does this PR solve?

waiting for leader change and avoid specific pd-name
test 90 times without error https://github.com/HuSharp/pd/pull/9

Issue Number: Close #8351

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]

File: tests/server/region_syncer/region_syncer_test.go [hunk 1]
```diff
@@ -282,16 +284,22 @@ func TestPrepareCheckerWithTransferLeader(t *testing.T) {
...
+	rc = cluster.GetLeaderServer().GetRaftCluster()
+	re.True(rc.IsPrepared())
 
-	// transfer leader to pd1, can start coordinator immediately.
+	// transfer leader, can start coordinator immediately.
+	leaderServer = cluster.GetLeaderServer()
 	err = cluster.ResignLeader()
 	re.NoError(err)
-	re.Equal("pd1", cluster.WaitLeader())
+	re.NotEqual(leaderServer.GetLeader().GetName(), cluster.WaitLeader())
```
Comment (rleungx) [inline tests/server/region_syncer/region_syncer_test.go:301]: What if the WaitLeader still return `""`?

Comment (HuSharp) [reply tests/server/region_syncer/region_syncer_test.go:301]: If we don't get a reasonable leader, maybe we can assume the test has failed?

Comment (rleungx) [reply tests/server/region_syncer/region_syncer_test.go:301]: The previous failure is caused by no leader and after this PR, will it still fail?

Comment (HuSharp) [reply tests/server/region_syncer/region_syncer_test.go:301]: I added a detailed explanation in 
https://github.com/tikv/pd/issues/8351#issuecomment-2205228978

I think in this extreme situation, we can fail directly.

Comment (HuSharp) [reply tests/server/region_syncer/region_syncer_test.go:301]: I created an issue to focus on Waitleader failure situation :) https://github.com/tikv/pd/issues/8359

**Review by rleungx — APPROVED** [review]

**Pull request merged**